### ISSUE TITLE
OutOfMemory w/100+GB free

### ISSUE THREAD

**Issue by MartinBrugnara** [arangodb/arangodb#1939]
## Issue:
ArangoDB crashed reporting the following lines, "out of memory".
Even tough there are 100+GB of free RAM

```
2016-07-12T21:17:30Z [1] INFO file-descriptors (nofiles) hard limit is 1048576, soft limit is 524288
2016-07-12T21:17:30Z [1] INFO Authentication is turned off
2016-07-12T21:17:30Z [1] INFO ArangoDB (version 2.8.9 [linux]) is ready for business. Have fun!

#
# Fatal error in Scavenge promotion failed
# Allocation failed - process out of memory
#
```

## Context:
Connecting w/gremlin as: 
```
c=new ArangoDBConfiguration("arango", 8529);
g=new ArangoDBGraph(c, "mysuperdb", "V", "E")
```

Collection `V` and `E` was previously imported with `arangoimp`.
Size on disc of the json file: 46GB

**Comment by dothebart** [comment]
Hi, 
This is an error message of the V8 garbage collector. Parts of the blueprints API had been implemented using Javascript. 

In general the future save and recommended way of working with Graphs in ArangoDB is to work with [pattern matching traversals](https://docs.arangodb.com/3.0/AQL/Graphs/Traversals.html) through AQL, which are implemented natively and thus are able to utilize all available resources of your machine.

**Comment by MartinBrugnara** [comment]
Hi @dothebart ,
Thank you for answer.

ArangoDB will not be the only graph database I'll to talk to...that's why I choose gremlin ...
There is no way to make it work without go native?

**Comment by dothebart** [comment]
Hi Martin, 
the current gremlin definition doesn't fit well to ArangoDBs internal data model - thus the blueprints wrapper layer available in ArangoDB 2.8. Similar to ORMs on top of classical relational databases you won't get the maximum out of your system if you use the Tinkerpop wrapper. 

We evolved the  ArangoDB graphing features a lot with the pattern matching traversals to give users more power over their data. The result is a very well readable syntax within AQL where you can freely combine document queries with graph queries - which enables you to effectively work with your data inside of ArangoDB and fetch the results with the ArangoDB Java Driver.

In the future we may see a better ArangoDB Tinkerpop3 implementation, but to be honest, I can't give you any ETA if and when this is going to be available. 

So, the answer currently is - yes, if you want to work with big graphs in ArangoDB you have to use AQL.

**Comment by MartinBrugnara** [comment]
Ok, got it!
Thank you so much!
